### PR TITLE
Additional git tagger variants (TreeSha, AbbrevTreeSha)

### LINKED ISSUE #407: Enhance git tagger to tag with latest git tag or other formats
The git tagger currently doesnt take any parameters but should take some to configure how we resolve  image tags from git history. Some options I'd like to see, not sure if they should be exposed as bool parameters or a string enum.

```yaml
# tags with the full commit sha256 instead of short
type: longCommit

# uses the most recent tag
type: tag
```

### PR DESCRIPTION
**Pull request by corneliusweig**
### Synopsis
PR #1902 has introduced git tagger variants `Tags`, `CommitSha`, and `AbbrevCommitSha`. Issue #407 also suggests tree hash variants, which are part of this PR. Originally these variants were also part of #1902, but OSX has problems to determine the relative git path, so these two variants were extracted.

- TreeSha: use the full tree hash of the artifact workingdir
- AbbrevTreeSha: use the abbreviated tree hash of the artifact workingdir

The tree hash variants are similar in spirit to the `sha256` tagger in that it is based on the (committed) workingdir content. As the hash is pre-computed by git, this should be faster than the `sha256Tagger`.

### Config change
There are no config changes wrt. PR #1902, but `gitCommit.variant` now understands the `TreeSha` and `AbbrevTreeSha` variants:
```yaml
tagPolicy:
  gitCommit:
   # variant is an enum with valid values (case-insensitive)
   # - Tags
   # - CommitSha
   # - AbbrevCommitSha
   # - TreeSha                  NEW
   # - AbbrevTreeSha       NEW
    variant: Tags # default
```
Close #407

### CODE REVIEW

**Comment by corneliusweig** [comment]
So the problem on OSX seems to be that `filepath.Abs()` reports `/var/folders/...` for the workingdir, but `git rev-parse --show-toplevel` prepends an additional folder: `/private/var/folders/...`.

> absWd=/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/skaffold810560671
> gitRoot=/private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/skaffold810560671

**Comment by nkubala** [comment]
> So the problem on OSX seems to be that filepath.Abs() reports /var/folders/... for the workingdir, but git rev-parse --show-toplevel prepends an additional folder: /private/var/folders/....

I actually don't see this when I run `git rev-parse --show-toplevel` on my machine (`git version 2.19.0.605`). might just be a weird quirk of some versions of git. in the case that it fails, is there something we can use as a fallback? maybe try and strip off the `/private` prefix if we get a failure, and then actually fail if that also doesn't work?

**Comment by corneliusweig** [comment]
@nkubala So I think the root cause is that `/var` on some OSX variants is actually a symlink to `/private/var`. Go `Abs` does not expand the symlink, whereas `git rev-parse --show-toplevel` does. At least this is my assumption, because in lack of an OSX system I can't be sure.
However, I would not put in some extra logic for stripping `/private`, this just feels horribly brittle. For example, if an affected OSX user has another symlink somewhere in his dir hierarchy, this would fail already.
Maybe you can do further checks on your machine?

File: pkg/skaffold/build/tag/git_commit.go [hunk 1]
```diff
@@ -94,13 +101,44 @@ func (c *GitCommit) makeGitTag(workingDir string) (string, error) {
...
+		return "", err
+	}
+
+	// git reports the gitdir with resolved symlinks, so we need to do this too in order for filepath.Rel to work
+	absWorkingDir, err = filepath.EvalSymlinks(absWorkingDir)
+	if err != nil {
+		return "", err
+	}
+
+	gitRoot, err := runGit(workingDir, "rev-parse", "--show-toplevel")
```
Comment (tejal29) [inline pkg/skaffold/build/tag/git_commit.go]: `runGit` is working as is in "GenerateFullyQualifiedImageName" function. 
Maybe resolving workspace should be part of ruunGit and not the other way around?

Comment (tejal29) [reply pkg/skaffold/build/tag/git_commit.go]: hmm looks like the working dir is something else than "git root"
is that correct?

Comment (corneliusweig) [reply pkg/skaffold/build/tag/git_commit.go]: @tejal29 `git root` is the top level directory of the git repository, whereas `workingDir` is the top level of the artifact to be tagged. I don't think this functionality can be moved into `runGit`, though. But maybe you have better ideas how to improve the variable naming?

Comment (tejal29) [reply pkg/skaffold/build/tag/git_commit.go]: maybe a comment for GenerateFullyQualifiedImageName could describe the workingDir is the artifact directory would be more clear?
```
// GenerateFullyQualifiedImageName tags an image with the supplied image name and the git commit.
func (c *GitCommit) GenerateFullyQualifiedImageName(workingDir string, imageName string) (string, error) {
```

Comment (corneliusweig) [reply pkg/skaffold/build/tag/git_commit.go]: @tejal29 I added some godoc to the `GenerateFullyQualifiedImageName` interface. Does that help?

Comment (tejal29) [reply pkg/skaffold/build/tag/git_commit.go]: Yes!

**Comment by dgageot** [comment]
@corneliusweig can this be merged?

**Comment by corneliusweig** [comment]
@dgageot No, it should come after #2109 (config freeze). Unless this change goes under the radar config-wise. All it does is add new enum values, so this would be possible without breaking anybody.

**Comment by corneliusweig** [comment]
@nkubala Done. I merged in master after the config freeze.

**Pull request merged**